FE_Enriched

davydden opened the pull request:
a new FE to do partition of unity method, where enriched and non-enriched elements can be combined.

- [x] squash
- [x] add an entry in `changes.h`
- [x] change `we store the casted result in` -> `we store the cast result in`
- [x] remove `Threads::Mutex`
- [x] remove destructor.
- [x] instantiate for `dim != spacedim`
- [x] require c++11
- [x] add a note on disjoint domains and different functions for different cells
- [x] redraw `doc/doxygen/images/fe_enriched_1d.png`
- [ ] if the accepted version of the paper describes implementation, add it to `publications.html`.
- [ ] consider writing a tutorial to show how to use the most complicated constructor with `std::function`.

davydden commented on tests/fe/fe_enriched_step-36.cc:
```diff
@@ -0,0 +1,948 @@
+          potential.value_list (fe_values.get_quadrature_points(),
+                                potential_values);
+
+#ifdef WITHOUT_FE_ENRICHED
```
for those interested, that's the part to look for a comparison between the brute force PUM and using FE_Enriched.

bangerth replied:
I think it would be useful to figure out a way how we can run this test with both variants -- because that would give us one more test. One way would be to clone it (including code duplication) and in each of the two `.cc` files delete one branch of the `#ifdef`, once you're satisfied that you've got the final version of the test. There are of course also other ways to avoid the code duplication.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+ *
+ * Note that it would not be wise to have this class derived
+ * from FESystem as the latter is a vector-valued FE, whereas the current class
+ * can be both scalar and vector valued depending on the base FEs used to build it.
```
I think it would be more comprehensive if you said "... as the latter concatenates the two given elements into different components of a vector element, whereas the current class combines the two given elements into the same components. If two scalar elements are given, the resulting element will be scalar rather than have two components when doing the same with an FESystem, for instance."

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+ * all enriched DoFs on the interface are constrained to be zero.
+ * This is a natural outcome of FE_Nothing with domination set to <code>true</code>.
+ *
+ * Note, that if the constraints are built between two different instances of this class,
```
remove comma after "Note"

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+ * Note, that if the constraints are built between two different instances of this class,
+ * no checks are performed to ensure that the underlying enrichment function is
+ * the same. Thus the user must be careful in order not to end up with the
+ * discontinuous FE space.
```
the discontinuous FE space -> a discontinuous FE space.

davydden replied:
fixed.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base,
+               const FiniteElement<dim,spacedim> &fe_enriched,
+               const Function<dim>      *enrichment_function);
```
Any reason why you expect to get a pointer rather than a function object (or reference)?

davydden replied:
that's because i use this one from the wrapper constructor `FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);` with `FE_Nothing` and `NULL` for the function.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+
+  /**
+   * Constructor which only wraps the base FE @p fe_base.
+   * As for enriched FiniteElement, FE_Nothing is used.
```
Say "As for the enriched element"

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);
+
+
```
remove one empty line, we typically only put one empty line between function declarations inside a class.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
+               const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched,
+               const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions);
```
I assume the nested `vector<vector<...>` construction is that you want to add different functions for different cells? Maybe add a more verbose discussion in the description of the constructor. Also, I'm not completely clear on how the enrichment is combined when having multiple vectors...

davydden replied:
correct, it's aimed for cases when you have disjoint enrichment domains with the same base and enriched FE(s) but different functions. I will add a discussion on this.

kronbichler commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+     * because this would mean that (i) only one FEValues object could use a finite element object at a time,
+     * and (ii) that these objects could not be used in a multithreaded context.
+     */
+    mutable std::vector<std::vector<EnrichmentValues>> enrichment;
```
Nice!

kronbichler commented on include/deal.II/fe/fe_system.h:
```diff
@@ -1071,6 +1072,9 @@ class FESystem : public FiniteElement<dim,spacedim>
    */
   mutable Threads::Mutex mutex;
+
+
```
Use only a single empty line.

luca-heltai commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,519 @@
+   * @p base_no_mult_local_enriched_dofs[i][j] contains the associated local DoFs
+   * on the FE_Enriched finite element.
+   */
+  std::vector<std::vector<std::vector<unsigned int>>> base_no_mult_local_enriched_dofs;
```
Here and everywhere else: `>>>` => `> > >`

luca-heltai replied:
The `>>` format is not compatible with `c++98`...

davydden replied:
this is c++11 class (use it heavily in constructors), i will add `ifndef`.

davydden commented:
It looks like github does not collapse comments on modified liens anymore. I addressed the issues pointed out so far.

bangerth commented:
Looking through this, if you still have the data for `doc/doxygen/images/fe_enriched_1d.png`, please regenerate it with larger axes labels and key -- otherwise it will not be possible to see what all of this is once you include the picture into the html.

bangerth commented on doc/doxygen/stylesheet.css:
```diff
@@ -26,3 +26,11 @@ div.doxygen-generated-exception-message {
 }
+
+div.image img[src="fe_enriched_1d.png"] {
+    width:500px;
```
oh, that seems like the wrong place to adjust the width (because it's going to be impossible to find if anyone ever wonders where the width is set, and because it will definitely be forgotten to remove this text if the figure is ever removed or renamed). Please attach this (and the next) width directly to the place where you have the `@img` tag.

davydden replied:
AFAIK what you ask is not possible, thus the workaround.

davydden replied:
i tried `@image html fe_enriched_1d.png width=500px` but it does not work.

davydden replied:
here is what they say https://www.star.bnl.gov/public/comp/sofi/doxygen/commands.html

> The fourth argument is also optional and can be used to specify the width or height of the image. This is only useful for $\mbox{\LaTeX}$ output (i.e. format=latex). The sizeindication can be either width or height. The size should be a valid size specifier in $\mbox{\LaTeX}$ (for example 10cm or 6in or a symbolic width like \textwidth).

davydden replied:
`@image html fe_enriched_1d.png "" width=500px` does not work either. So i guess we need to live with this solution until Doxygen gets more flexible.

bangerth replied:
Hm, bummer. What if you just saved the image in a size that is already roughly correct? Alternatively, just put a comment in front of the `@image` tag that explains where the size is set. (Which raises the question of how to put a comment into doxygen -- I don't really know...)

davydden replied:
i will keep it as is, unless you have a strong opinion on the image size issue. 

Generally i think it is good to have higher resolution images than the size they are displayed. At the end of the day, that's exactly what CSS is for. 

As for knowing who controls the size, it's easy to find out with [Firebug](https://addons.mozilla.org/en-US/firefox/addon/firebug/) plugin to Firefox. Very handy for web development.

bangerth replied:
OK.

bangerth commented on include/deal.II/fe/fe.h:
```diff
@@ -2722,6 +2723,7 @@ class FiniteElement : public Subscriptor,
   friend class FEFaceValues<dim,spacedim>;
   friend class FESubfaceValues<dim,spacedim>;
   friend class FESystem<dim,spacedim>;
+  friend class FE_Enriched<dim,spacedim>;
```
Out of curiosity, why do you need this?

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+  /**
+   * Thread mutex for protecting multiply_by_enrichment().
+   */
+  mutable Threads::Mutex mutex;
```
I don't think I understand why you need this. The function is `const`, all of its input arguments are `const`, and there are no other `mutable` variables in the class. Consequently, there is nothing other than the mutable members of the `fe_internal` object that you could possibly modify, and that object is not accessed from different threads anyway. So I think you should be ok without the mutex.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ * \f[
+ * U(x) = \sum_i N_i(x) U_i + \sum_j N_j(x) \sum_k F_k(x) U_{jk}
+ * \f]
+ * where \f$ N_i(x) \f$ and \f$ N_j(x) \f$ are the underlying finite elements (including
```
No need for `\f` -- we automatically expand the dollar sign to `\f$`.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ * \f]
+ * where \f$ N_i(x) \f$ and \f$ N_j(x) \f$ are the underlying finite elements (including
+ * the mapping from the isoparametric element to the real element); \f$ F_k(x) \f$
+ * are the enrichment functions in real space (e.g. \f$ 1/r \f$, \f$ exp(-r) \f$, Heaviside, etc);
```
-> `\exp{-r}`

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ * where \f$ N_i(x) \f$ and \f$ N_j(x) \f$ are the underlying finite elements (including
+ * the mapping from the isoparametric element to the real element); \f$ F_k(x) \f$
+ * are the enrichment functions in real space (e.g. \f$ 1/r \f$, \f$ exp(-r) \f$, Heaviside, etc);
+ * \f$ U_i \f$ and \f$ U_{jk} \f$ are the standard and enriched DoFs.
```
Does this mean that the class declares degrees of freedom for all of these U_{i}, U_{jk}? Can you describe what kind of object each of these DoFs are located on, and what their total number is? For the latter step, it may be useful to provide upper bounds to the sum above.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ *                     function)
+ * @endcode
+ *
+ * The \f$ exp(-x) \f$ enrichment leads to the following shape functions on unit
```
same as above

bangerth replied:
when you say "the ... enrichment", I assume you mean that a function that returns this term is used in place of `function` in the argument list above?

davydden replied:
> when you say "the ... enrichment", I assume you mean that a function that returns this term is used in place of function in the argument list above?

correct.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ *   </tr>
+ *   <tr>
+ *     <td align="center">
+ *       1d element, base and partition-of-unity shape functions.
```
by the way, why is it called "partition of unity"? I think it would be useful to explain this term somewhere towards the top of the documentation.

davydden replied:
pum

davydden replied:
done

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ *       1d element, base and partition-of-unity shape functions.
+ *     </td>
+ *     <td align="center">
+ *       enriched shape function resulting from h-refinement of a single element (bottom-left corner).
```
maybe say "enriched shape function corresponding to the central vertex"?

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ *   </tr>
+ * </table>
+ *
+ * <h3>Implementation</h3>
```
I think this should be at the very bottom of the documentation, as it is not really of interest to someone who just wants to *use* the class.

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ * This class implements hp constraints with arbitrary FE. For the example above,
+ * a user will need to wrap a given FiniteElement into FE_Enriched:
+ * @code
+ *   FE_Enriched<dim> fe_enriched(FE_Q<dim>(1));
```
I don't think I understand this. `FE_Q` already implements hp constraints, so why does one have to wrap anything? And what does it mean to wrap `FE_Q` into an `FE_Enriched` without any enrichment functions?

davydden replied:
because you can't have constraints between `FESystem(FE_Q,FE_Q)` and `FE_Q`, but you can between `FESystem(FE_Q,FE_Q)` and `FESystem(FE_Q,FE_Nothing)`. That's why you need to wrap non-erniched elements into `FE_Enriched`.

bangerth replied:
Oh, so are you saying that one would use the wrapper because this allows to use `FE_Enriched(FE_Q)` on one cell, and `FE_Enriched(FE_Q, FE_Q, function)` on another and automatically get constraints working, whereas the same would not be true for `FE_Q` and `FE_Enriched(FE_Q, FE_Q, function)`?

If this is so, then I think this is unnecessarily complicated. Just force users to use `FE_Enriched (FE_Q, FE_Nothing, something)` instead. This saves you a bunch of code and makes things a lot simpler. Alternatively, we may be able to deal with this internally in the functions that compute the constraints.

davydden replied:
> Oh, so are you saying that one would use the wrapper because this allows to use FE_Enriched(FE_Q) on one cell, and FE_Enriched(FE_Q, FE_Q, function) on another and automatically get constraints working, whereas the same would not be true for FE_Q and FE_Enriched(FE_Q, FE_Q, function)?

exactly! The point is to reuse `FESystem` as much as possible.

> If this is so, then I think this is unnecessarily complicated. Just force users to use FE_Enriched (FE_Q, FE_Nothing, something) instead. This saves you a bunch of code and makes things a lot simpler. 

It could make more difficult for users as they need to be aware to use `FE_Nothing(dominate)` to get constraints working. But I can of course add checks to constructors that this is the case.

On another hand, i agree with your that having two constructors `FE_Enriched(FE_Q, FE_Q, function)` and the complicated general one below is more logical.

bangerth replied:
After thinking of this some more, I think I would be ok with this constructor if (i) the introduction to this class clearly explained *why* one would want to wrap an element, (ii) explained what this does internally (namely using the `FE_Nothing` as you explain), and (iii) the constructor's documentation referenced this discussion in the class documentation.

I think you have a good point, but the documentation so far doesn't help connect these dots and explain. If you packaged up the discussion we're having here into the class doc, that would already be sufficient.

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ *
+ * @code
+ * FESystem<dim>(FE_Q<dim>(1),1,
+ *               FE_Nothing<dim>(1,true),1)
```
I don't think I understand this either. The `FE_Enriched` is scalar, whereas this one is not. How can you build constraints?

davydden replied:
constraints are borrowed from FESystem. So constrants make sure that each component of two FESystems are continuous. This in turn will make sure that scalar FE space produced by FEEnriched (with and without enrichment) is also continuous.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+ * @endcode
+ *
+ * In order to make the resulting field continuous
+ * all enriched DoFs on the interface are constrained to be zero.
```
at the interface of what?

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   * this non-enriched element is used in conjunction with enriched finite element
+   * within the hp::DoFHandler.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);
```
I don't think I understand the documentation. What kinds of basis functions does this element actually have? If you enrich with `FE_Nothing`, then there should be no additional basis functions because `FE_Nothing` has no basis functions.

davydden replied:
correct. this is a wrapper for non-enriched element.

bangerth replied:
So then it's not clear to me why anyone would want to use this. What's the functionality of a wrapped element without enrichment?

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+  /**
+   * Constructor which takes pointer to the base FiniteElement @p fe_base and
+   * a vector of enriched FiniteElement's @p fe_enriched . @p i'th element of this
+   * vector will be enriched with @p i'th functions in @p functions.
```
I think this could use a formula :-)

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+
+  /**
+   * Access to a composing element. The index needs to be smaller than the
+   * number of base elements.
```
can you document what "base element" means in the current context?

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   * Return the value of the @p ith shape function at the point @p p. @p p is a
+   * point on the reference element.
+   *
+   * This function returns meaningful value only for non-enriched element as
```
value -> values

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
+               const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched,
+               const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions);
```
Would it be simpler to use a `Table<2,std::function<...>>`?

davydden replied:
i don't think so, see implementation of constructors.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   */
+  struct EnrichmentValues
+  {
+    std::vector<double> values;
```
Since you only need this structure in `InternalData`, maybe move it there?

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+    get_fe_output_object (const unsigned int base_no) const;
+
+    /**
+     * Aggregate FESystem's internal data. It is used it every time
```
-> it is used

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+{
+  typename FESystem<dim,spacedim>::InternalData *data_fesystem =
+    static_cast<typename FESystem<dim,spacedim>::InternalData *> (fes_data);
+  Assert (data_fesystem != NULL,
```
I think you want a `dynamic_cast` here

davydden replied:
i guess i can stick with `static_cast` and remove `Assert` because this function is called only inside the class always with `fe_system.get_face_data()` and alike.

bangerth replied:
That's one way. In all of the other FE classes, I believe that we check with `dynamic_cast` inside the `Assert` that the type is really correct, and then do a `static_cast` as you do. In other words, just swap the `Assert` and the `static_cast` you have here, and do a `dynamic_cast` inside the `Assert`.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+     * As we aggregate FESystem's internal data, we only call it after
+     * the FESystem::get_data() is called.
+     */
+    InternalData ( typename FESystem<dim,spacedim>::InternalData *fesystem_data);
```
If I understand things correctly here, the current object will take over ownership of the object pointed to. I think you could make this clearer by making the argument a `std_cxx11::unique_ptr` and the same for the member variable of the current class. This way you have a nice chain of passing ownership from the calling site (where you would use `std_cxx11::make_unique`) to the place where the object is finally stored.

davydden replied:
i don't think i can do that. look at `setup_data` function, which takes 
```
typename FiniteElement<dim,spacedim>::InternalDataBase *fes_data
``` 
from `fe_system.get_face_data(update_flags,mapping,quadrature,output_data)`, then this is casted into 
```
typename FESystem<dim,spacedim>::InternalData *data_fesystem =
    static_cast<typename FESystem<dim,spacedim>::InternalData *> (fes_data);
```.
Finally i setup some things with it.

davydden replied:
nevermind, i think it is perfectly possible, i mixed up the logic. Will try it now...

davydden replied:
I updated the documentation of constructor `InternalData` and also `FE_Enriched::setup_data`. Now i think i understand again what actually i was doing here :smile:.

p.s. IIRC `FE_Enriched::setup_data()` got unnecessary complicated because i was trying to implement another variant of the method, but later decided to cut it off.

davydden replied:
oh, i see now. I updated the code to `unique_ptr`.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+    /**
+     * Destructor. Deletes @p fesystem_data pointer.
+     */
+    ~InternalData();
```
You could then also get rid of this here, because the `unique_ptr` takes care of it

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+  public:
+    /**
+     * Constructor. Is called by the @p get_data function.
+     * As we aggregate FESystem's internal data, we only call it after
```
what is "it" here?

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+    get_fe_data (const unsigned int base_no) const;
+
+    /**
+     * Gives read-access to the pointer to an object to which into which the
```
grammar: to which into which

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+     * because this would mean that (i) only one FEValues object could use a finite element object at a time,
+     * and (ii) that these objects could not be used in a multithreaded context.
+     */
+    mutable std::vector<std::vector<EnrichmentValues> > enrichment;
```
could you convert this into a `Table<2,EnrichmentValues>`?

davydden replied:
i don't think so because each FE used with enrichment can be coupled with an arbitrary number of enrichment functions. That is, you may have
```
FE_Q(2)  // base FE
FE_Q(1)  + sin(x) + cos(x) // first element with enrichment
FE_Q(2) + exp(-x) // second element with enrichment
```

bangerth replied:
I see.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   * @p base_no_mult_local_enriched_dofs[i][j] contains the associated local DoFs
+   * on the FE_Enriched finite element.
+   */
+  std::vector<std::vector<std::vector<unsigned int> > > base_no_mult_local_enriched_dofs;
```
Could this be made into a `Table<2,vector<uint>>` or even `Table<3,uint>`?

davydden replied:
no, see https://github.com/dealii/dealii/pull/3115/files#r80472010

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   * Auxiliary variable used to distinguish between the case when we do enrichment
+   * and when the class simply wraps another FiniteElement.
+   */
+  const bool is_enriched;
```
It's not clear to me why one would want to wrap an element in the first place, but aren't there ways to compute whether the element is enriched from other member variables? It seems to me like you're storing duplicate information here, which is always prone to the introduction of bugs later on.

If the expression to compute whether the element is enriched from other members is complicated, simply introduce a function `is_enriched()` that encapsulates the expression.

davydden replied:
> aren't there ways to compute whether the element is enriched from other member variables

yes, exactly what is done in the constructor: loop over a vector of enrichment elements and check if all of them are FE_Nothing. I don't think this should be done every single time (i.e. on every element) we need this info. Moreover, the member is `const`, so it is initialized once in constructor and there is now way to mess up with it.

> It seems to me like you're storing duplicate information here, which is always prone to the introduction of bugs later on.

in this case i think it appropriate solution.

bangerth replied:
OK. Can you convert your answer into a doc comment for this variable?

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   *
+   * This function is needed by the constructors of @p FESystem.
+   */
+  virtual FiniteElement<dim,spacedim> *clone() const;
```
I think this is a `public` member function in the base class, so maybe move it up to the `public` section here as well?

davydden replied:
done.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,545 @@
+   */
+  template <int dim_1>
+  typename FiniteElement<dim,spacedim>::InternalDataBase *
+  setup_data (typename FiniteElement<dim,spacedim>::InternalDataBase *fes_data,
```
can you explain what the arguments are, and in particular what the pointer argument in first place here is (and who has ownership)?

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+  if (const FE_Enriched<dim,spacedim> *fe_enr_other
+      = dynamic_cast<const FE_Enriched<dim,spacedim>*>(&source))
+    {
+      fe_system.get_subface_interpolation_matrix(fe_enr_other->get_fe_system(),
```
ah, this is quite neat. I think I finally understand why you base everything on an fe system :-)

bangerth commented on source/fe/fe_enriched.inst.in:
```diff
@@ -0,0 +1,22 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2015 by the deal.II authors
```
adjust years

bangerth commented on source/fe/fe_enriched.inst.in:
```diff
@@ -0,0 +1,22 @@
+
+for (deal_II_dimension : DIMENSIONS)
+{
+    template class FE_Enriched<deal_II_dimension, deal_II_dimension>;
```
any particular reason why you don't instantiate for `dim != spacedim`?

bangerth replied:
I see no reason why this should not work.

bangerth commented on tests/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,794 @@
+// ---------------------------------------------------------------------
+
+
+// basic checks of FE_Enriched class (see description for each small subtest)
```
This is a *really* long test. If something goes wrong, it will be quite difficult to see right away which part of the test it is that broke. Consequently, we have gradually moved away from such large tests and just split them into many smaller ones that all only test one individual aspect of a class. If they share significant code, you can always put the shared code (and/or `main` function) into a `.h` file that you `#include` from all of the tests. There are numerous examples of this style in the testsuite.

bangerth commented on tests/fe/fe_enriched_step-36.cc:
I'm not a fan of this. I would say that debug and release mode tests should run the same code/number of iterations

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+  {
+    [=] (const typename Triangulation<dim, spacedim>::cell_iterator &) -> const Function<dim> *
+    {
+      return enrichment_function;
```
done.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+
+
+template <int dim, int spacedim>
+FE_Enriched<dim,spacedim>::FE_Enriched (const FiniteElement<dim,spacedim> &fe_base)
```
done.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+  // note that the order if important, namely
+  // output_data.shape_XYZ contains values of standard FEM and we overwrite
+  // it with the updated one in the following order: hessians -> gradients -> values
+  if (flags & update_hessians)
```
done.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+template <int dim, int spacedim>
+FE_Enriched<dim,spacedim>::~FE_Enriched ()
+{
+}
```
since the class has virtual functions, don't we need to provide virtual destructor? Not necessarily someone will derive anything from `FE_Enriched` though.

bangerth replied:
Because the destructor of the base class is `virtual`, the one of the current class is also `virtual` -- whether you actually declare/define it or not.

davydden commented:
@bangerth I addressed all your comments but the one related to passing FE as a pointer or reference. Will look at it later. Should be ready for the next round of review.

ps. thanks for reviewing, it definitely increases the quality of the code.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+ * enriched element is the sum of DoFs from <code>FE_Q<dim>(2)</code> and
+ * <code>FE_Q<dim>(1)</code>.
+ *
+ * As an example of enrichment function, consider $ \exp(-x) $, which
```
example on AN enrichment function

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+ * <code>FE_Q<dim>(1)</code>.
+ *
+ * As an example of enrichment function, consider $ \exp(-x) $, which
+ * leads to the following shape functions on unit element:
```
on THE unit element

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+ * In most applications it is beneficial to introduce enrichments only in
+ * some part of the domain (e.g. around a crack tip) and use standard FE (e.g. FE_Q)
+ * elsewhere. In order to make the resulting FE space $C^0$ continuous in the
+ * case of continuous enrichmenet functions, all enriched DoFs on the interface
```
-> enrichment

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+ *                                    NULL);
+ * @endcode
+ * and will result in the correct constraints for enriched DoFs attributed to
+ * support points on the interface between the two regions.
```
I think you want to mention the word "hp" here somewhere. Maybe something like this after the first sentence "...elsewhere.":
```
  This can be achieved by using the hp finite element framework in deal.II that allows for the use of different
  elements on different cells. To make the resulting space $C^0$ continuous, it is then necessary for
  the DoFHandler class and DoFTools::make_hanging_node_constraints() function to be able to figure
  out what to do at the interface between enriched and non-enriched cells. Specifically, we want the
  degrees of freedom corresponding to enriched shape functions to be zero at these interfaces. These
  classes and functions can not to do this automatically, but the effect can be achieved by using not
  just a regular FE_Q on cells without enrichment, but to wrap the FE_Q into an FE_Enriched object
  <i>without actually enriching it</i>. This can be done as follows:
```
and then just continue with the code example.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+ * The ordering of the shape function, @p interface_constrains, the @p prolongation (embedding)
+ * and the @p restriction matrices are taken from the FESystem class.
+ *
+ * <h3>References</h3>
```
I would move that even above the "Implementation" section.

Does the paper actually discuss implementation details, or just show results? In the former case, we may want to consider listing it along with the other papers at the top of `publications.html` that discuss implementations.

davydden replied:
> Does the paper actually discuss implementation details, or just show results?

both. But it is the very first submitted version, so hopefully implementation details will survive the review. We wait 7 months already...

bangerth replied:
OK. All I want to say is that if the paper sheds light on a piece of implementation of deal.II, then feel free to put it on the list (or ask me to put it on the list) once published.

davydden replied:
will do once it is (hopefully) accepted.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+}
+@Article{Babuska1997,
+  Title                    = {The partition of unity method},
+  Author                   = {Babu\v{s}ka, I. and Melenk, J. M.},
```
Just a minor nitpick, all three of these references differ in the way they abbreviate (or not) first names, and where they appear relative to last names.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+   * this non-enriched element is used in conjunction with enriched finite element
+   * within the hp::DoFHandler.
+   *
+   * See the discussion in the class documentation.
```
...on how to use this element in the context of hp finite element methods.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
+               const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched,
+               const std::vector<std::vector<std::function<const Function<spacedim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions);
```
the signature of the last argument is quite complex. I think it would be useful to have an example that explains how to use it.

Or did you intend to write a little tutorial program, maybe, that shows this? That would be even better!

davydden replied:
> Or did you intend to write a little tutorial program, maybe, that shows this? That would be even better!

i have not thought about this, but i can think of some example which would illustrate the usage case.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+     * data object. The former is called from get_data, get_subface_data and
+     * get_face_data which FE_Enriched has to implement.
+     *
+     * Since FESystem::get_face_data() just create an object and return a
```
create -> creates
return -> returns

It's also get_data() and get_subface_data(), btw.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+     * get_face_data which FE_Enriched has to implement.
+     *
+     * Since FESystem::get_face_data() just create an object and return a
+     * pointer to it (i.e. hey don't retain ownership), we store the casted
```
hey?

bangerth replied:
the past tense of "cast" is still "cast"

davydden replied:
that's an adjective :smile:

davydden replied:
of course you are right, `casted` is completely different thing.

davydden replied:
although in some places is that `casted` is 

> simple past tense and past participle of cast

@drwells : is it `we store the casted result in...` or `we store the cast result in...`?

davydden replied:
fixed.

bangerth commented on include/deal.II/fe/fe_system.h:
```diff
@@ -1071,6 +1072,8 @@ class FESystem : public FiniteElement<dim,spacedim>
    */
   mutable Threads::Mutex mutex;
+
+  friend class FE_Enriched<dim,spacedim>;
```
remind me why you needed this?

davydden replied:
because i use non-public functions of FESystem.

bangerth replied:
Sure, but which ones?

davydden replied:
fe_system.interface_constraints;

davydden replied:
`fe_system.unit_support_points` and `fe_system.unit_face_support_points` and `fe_system.adjust_line_dof_index_for_line_orientation_table`.

bangerth replied:
OK. That's enough to warrant the `friend` declaration.

bangerth commented on include/deal.II/fe/fe_enriched.h:
```diff
@@ -0,0 +1,585 @@
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_CXX11
```
Do you want to mention somewhere in the documentation that the element is only usable with C++11?

davydden replied:
done.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+  bool
+  consistency_check (const std::vector< const FiniteElement< dim, spacedim > * > &fes,
+                     const std::vector< unsigned int > &multiplicities,
+                     const std::vector<std::vector<std::function<const Function<spacedim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions)
```
It would be lovely if this function could check that the enrichment functions are really all scalar, because that's really the only case you support right now. But you can't do that here because you no longer have access to the `Function` objects without picking any particular cell.

I'm going to comment on this again below in `fill_fe_values`.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                                          const Quadrature<dim-1> &quadrature,
+                                          internal::FEValues::FiniteElementRelatedData< dim, spacedim >        &output_data) const
+{
+  return setup_data(std::unique_ptr<typename FiniteElement<dim,spacedim>::InternalDataBase>(fe_system.get_face_data(update_flags,mapping,quadrature,output_data)),
```
I don't think you need the explicit cast to `std::unique_ptr` here. The compiler should be able to do that for you.

bangerth replied:
Sure, but what I meant is to just write
```
  return setup_data (fe_system.get_face_data(update_flags,mapping,quadrature,output_data),
                                ...);
```

davydden replied:
that's not possible
```
/Users/davydden/libs-sources/deal.ii/davydden/include/deal.II/fe/fe_enriched.h:498:3: note: candidate function [with dim_1 = 2] not viable: no known conversion from 'typename FiniteElement<3, 3>::InternalDataBase *' to
      'std::unique_ptr<typename FiniteElement<3, 3>::InternalDataBase>' for 1st argument
  setup_data (std::unique_ptr<typename FiniteElement<dim,spacedim>::InternalDataBase> fes_data,
  ^
```

bangerth replied:
Ah, interesting. The constructor of that class must be `explicit` then. Oh well, thanks for trying!

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                                             const Quadrature<dim-1> &quadrature,
+                                             dealii::internal::FEValues::FiniteElementRelatedData<dim, spacedim> &output_data) const
+{
+  return setup_data(std::unique_ptr<typename FiniteElement<dim,spacedim>::InternalDataBase>(fe_system.get_subface_data(update_flags,mapping,quadrature,output_data)),
```
same here and below, of course

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                                            *this,
+                                            false);
+
+  }
```
remove empty line

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+  const UpdateFlags flags = fe_data.update_each;
+//  const UpdateFlags flags(dim_1==dim ?
+//                          fe_internal.current_update_flags() :
+//                          fe_internal.update_flags);
```
do you still need these commented out line?

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                      output_data.shape_hessians[out_index+s][q] =
+                        base_data.shape_hessians[in_index+s][q];
+
+                }
```
remove empty lines

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].hessians.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].hessians[q]  = enrichments[base_no-1][m](cell)->hessian (mapping_data.quadrature_points[q]);
```
here and in the same lines in the following two blocks, you call `enrichments[...][m](cell)` which gives you a `Function<dim>*`. I would like it if there was a check somewhere that this is really a scalar function. E.g., at the top of the loop over `m`, you could have
```
  Assert (enrichments[base_no-1][m](cell)->n_components() == 1, ...);
```

davydden replied:
done.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    grad_grad =  outer_product(output_data.shape_gradients[enriched_dof][q],  fe_data.enrichment[base_no][m].gradients[q]);
+                    sym_grad_grad = symmetrize (grad_grad) * 2.0; // symmetrize does [s+s^T]/2
```
is it obvious why you actually need to symmetrize this?

davydden replied:
i would say it is not, but last time i did derivations that was the outcome of the chain rule

bangerth replied:
I see. Might be worth expanding on in the documentation here or in the class documentation. In fact, writing out all of those (first and second) derivatives may be useful in the class doc because it would be useful to state there anyway that the `Function` object you pass in needs to be able to provide first and second derivatives if you want first or second derivatives of the enriched finite element.

davydden replied:
done.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+    {
+      Tensor<2, spacedim> grad_grad, grad_grad_r;
+      Tensor<2,spacedim,double> sym_grad_grad, sym_grad_grad_r;
+      Tensor<1,spacedim> grad_r;
```
move these into the innermost loops where you actually use them; you can then make them `const` in some of these places.

It probably doesn't matter very much, but I always like limiting the scope of variables because that makes understanding code so much simpler. I think it also makes it easier on the compiler because it has fewer live variables that need registers or stack space.

davydden replied:
good catch, extra unused variables were left-overs from another variant of the PUM.

bangerth commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,927 @@
+          for (unsigned int m=0; m < base_no_mult_local_enriched_dofs[base_no].size(); m++)
+            for (unsigned int i=0; i < base_no_mult_local_enriched_dofs[base_no][m].size(); i++)
+              {
+                const unsigned int &enriched_dof = base_no_mult_local_enriched_dofs[base_no][m][i];
```
no need to make this a reference, just leave it as a value. same below.

davydden commented:
the comments are addressed. `cast` vs `casted` is added to the TODO list in OP.

davydden commented on source/fe/fe_enriched.cc:
```diff
@@ -0,0 +1,949 @@
+
+
+template <int dim, int spacedim>
+FE_Enriched<dim,spacedim>::FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
```
One could do that, but i would still need to have `fe_enriched` to be a vector of pointers because eventually i call the most general constructor of `FESystem` with `const std::vector<const FiniteElement<dim,spacedim>*> &fes,`. 

To make the two input arguments consistent, i would prefer to keep `fe_base` as a pointer.

bangerth replied:
Yes, you need the vector of pointers. The situation is really like `FESystem`: It takes individual finite elements by reference, but a vector of pointers.

davydden commented:
I think i addressed all comments and did everything I planed to do with this PR. All its unit tests pass for me:
```
100% tests passed, 0 tests failed out of 26
```

bangerth commented:
Great, thanks! Nice feature. This needs to be mentioned in the next release notes!
